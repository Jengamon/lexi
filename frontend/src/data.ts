// Keep in sync with definitions from the Tauri app's data.rs
// TODO Figure out how to automate this?

import {
    Array,
    String,
    Record,
    Static,
    Union,
    Literal,
    Boolean,
    Unknown,
    Dictionary,
} from "runtypes";

export const PlosivePlace = Union(
    Literal("Bilabial"),
    Literal("Labiodental"),
    Literal("Dental"),
    Literal("Alveolar"),
);
export const FricativePlace = Union(
    Literal("Bilabial"),
    Literal("Labiodental"),
    Literal("Dental"),
    Literal("Alveolar"),
    Literal("Postalveolar"),
);
export const ObstruentAttachment = Union(
    Literal("Ejective"),
    Literal("Preaspirated"),
    Literal("Aspirated"),
    Literal("Breathy"),
    Literal("Creaky"),
);
export const Phone = Union(
    Record({
        Plosive: Record({
            place: PlosivePlace,
            voiced: Boolean,
            attachments: Array(ObstruentAttachment),
        }),
    }),
    Record({
        Affricative: Record({
            start_place: PlosivePlace,
            end_place: FricativePlace,
            voiced: Boolean,
            attachments: Array(ObstruentAttachment),
        }),
    }),
    Literal("Null"),
);
export const Phoneme = Record({
    ortho: String,
    primary: Phone,
    allo: Array(Phone),
});
export const Language = Record({
    name: String,
    description: Unknown.optional(),
    phonemes: Dictionary(Phoneme, String),
    ancestors: Array(String),
});
export const Protolanguage = Record({
    name: String,
    description: Unknown.optional(),
    phonemes: Dictionary(Phoneme, String),
});
export const LanguageGroup = Record({
    version: String,
    protolangs: Array(Protolanguage),
    langs: Array(Language),
});

export type PlosivePlace = Static<typeof PlosivePlace>;
export type FricativePlace = Static<typeof FricativePlace>;
export type ObstruentAttachment = Static<typeof ObstruentAttachment>;
export type Phone = Static<typeof Phone>;
export type Phoneme = Static<typeof Phoneme>;
export type Protolanguage = Static<typeof Protolanguage>;
export type Language = Static<typeof Language>;
export type LanguageGroup = Static<typeof LanguageGroup>;
