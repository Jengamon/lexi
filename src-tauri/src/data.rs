use semver::Version;
use std::borrow::Cow;
use std::collections::HashSet;
use std::fmt;
use std::fmt::Formatter;

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct LanguageGroup {
    pub version: Version,
    pub protolangs: Vec<Protolanguage>,
    pub langs: Vec<Language>,
}

// Use the crate version as the data version of the language group

// Since we use semver, we can just bake in our semver compatibility for data.
const CRATE_VERSION: &str = env!("CARGO_PKG_VERSION");

impl Default for LanguageGroup {
    fn default() -> Self {
        Self {
            version: Version::parse(CRATE_VERSION).unwrap(),
            protolangs: vec![],
            langs: vec![],
        }
    }
}

#[derive(Serialize, Deserialize, Clone)]
pub struct Protolanguage {
    pub name: String,
    pub phonemes: Vec<Phoneme>,
}

impl Default for Protolanguage {
    fn default() -> Self {
        Self {
            name: "unnamed".to_string(),
            phonemes: vec![],
        }
    }
}

#[derive(Serialize, Deserialize, Clone)]
pub struct Language {
    pub name: String,
    pub phonemes: Vec<Phoneme>,
}

impl Default for Language {
    fn default() -> Self {
        Self {
            name: "unnamed".to_string(),
            phonemes: vec![],
        }
    }
}

#[derive(Serialize, Deserialize, Clone)]
pub struct Phoneme {
    ortho: String,
    primary: Phone,
    allo: Vec<Phone>,
}

#[derive(Serialize, Deserialize, Copy, Clone)]
pub enum Place {
    Bilabial,
    Dental,
    Alveolar,
}

#[derive(Serialize, Deserialize, Hash, PartialEq, Eq, Copy, Clone)]
pub enum ObstruentAttachment {
    Ejective,
    Preaspirated,
    Aspirated,
    Breathy,
    Creaky,
}

enum PhoneType {
    Plosive,
    Fricative,
    Vowel,
}

#[derive(Serialize, Deserialize, Clone)]
pub enum Phone {
    Plosive {
        place: Place,
        voiced: bool,
        attachments: HashSet<ObstruentAttachment>,
    },
    Affricative {
        start_place: Place,
        end_place: Place,
        voiced: bool,
        attachments: HashSet<ObstruentAttachment>,
    },
    Fricative {
        place: Place,
        voiced: bool,
        attachments: HashSet<ObstruentAttachment>,
    },
    Vowel {},
}

impl Phone {
    fn get_base(
        pt: PhoneType,
        place: &Place,
        voiced: bool,
        attachments: &HashSet<ObstruentAttachment>,
    ) -> String {
        match pt {
            PhoneType::Plosive => match place {
                Place::Bilabial if attachments.contains(&ObstruentAttachment::Ejective) => {
                    "p".to_string()
                }
                Place::Bilabial if voiced => "b".to_string(),
                Place::Bilabial => "p".to_string(),
                Place::Dental if attachments.contains(&ObstruentAttachment::Ejective) => {
                    "t[".to_string()
                }
                Place::Dental if voiced => "d[".to_string(),
                Place::Dental => "t[".to_string(),
                Place::Alveolar if attachments.contains(&ObstruentAttachment::Ejective) => {
                    "t".to_string()
                }
                Place::Alveolar if voiced => "d".to_string(),
                Place::Alveolar => "t".to_string(),
            },
            PhoneType::Fricative => match place {
                Place::Bilabial if attachments.contains(&ObstruentAttachment::Ejective) => {
                    "P\"".to_string()
                }
                Place::Bilabial if voiced => "B\"".to_string(),
                Place::Bilabial => "P\"".to_string(),
                Place::Dental if attachments.contains(&ObstruentAttachment::Ejective) => {
                    "O-".to_string()
                }
                Place::Dental if voiced => "d-".to_string(),
                Place::Dental => "O-".to_string(),
                Place::Alveolar if attachments.contains(&ObstruentAttachment::Ejective) => {
                    "s".to_string()
                }
                Place::Alveolar if voiced => "z".to_string(),
                Place::Alveolar => "s".to_string(),
            },
            PhoneType::Vowel => todo!(),
        }
    }

    fn aspiration<'a>(phone: &'a str, attachments: &HashSet<ObstruentAttachment>) -> Cow<'a, str> {
        if attachments.contains(&ObstruentAttachment::Preaspirated) {
            Cow::Owned("h^".to_string() + phone)
        } else if attachments.contains(&ObstruentAttachment::Aspirated) {
            Cow::Owned(phone.to_string() + "h^")
        } else {
            Cow::Borrowed(phone)
        }
    }
}

impl fmt::Display for Phone {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let phone = match self {
            Phone::Plosive {
                place,
                voiced,
                attachments,
            } => {
                let mut phone = Self::get_base(PhoneType::Plosive, place, *voiced, attachments);

                if attachments.contains(&ObstruentAttachment::Breathy) {
                    phone += r#"h")"#;
                }

                if attachments.contains(&ObstruentAttachment::Creaky) {
                    phone += "~";
                }

                if attachments.contains(&ObstruentAttachment::Ejective) {
                    phone += "`";
                }

                phone = Self::aspiration(&phone, &attachments).to_string();

                phone
            }
            Phone::Affricative {
                start_place,
                end_place,
                voiced,
                attachments,
            } => {
                let mut start_phone =
                    Self::get_base(PhoneType::Plosive, start_place, *voiced, attachments);
                let mut end_phone =
                    Self::get_base(PhoneType::Fricative, end_place, *voiced, attachments);

                if attachments.contains(&ObstruentAttachment::Breathy) {
                    start_phone += r#"h")"#;
                    end_phone += r#"h")"#;
                }

                if attachments.contains(&ObstruentAttachment::Creaky) {
                    start_phone += "~\u{200b}";
                    end_phone += "~\u{200b}";
                }

                let mut phone = format!("{start_phone})){end_phone}");

                if attachments.contains(&ObstruentAttachment::Ejective) {
                    phone += "`";
                }

                phone = Self::aspiration(&phone, &attachments).to_string();

                phone
            }
            Phone::Fricative {
                place,
                voiced,
                attachments,
            } => todo!(),
            Phone::Vowel {} => todo!(),
        };

        write!(f, "{}", phone)
    }
}
