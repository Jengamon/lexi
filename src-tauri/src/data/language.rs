use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::Phoneme;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Language {
    /// Uniquely identifies this language
    ///
    /// Immutable once a language has been created.
    pub name: String,
    pub phonemes: HashMap<Uuid, Phoneme>,
    /// General information about the language
    ///
    /// Technically can be any JSON-able object
    pub description: Option<serde_json::Value>,
    /// Protolanguage names that this language is descended from
    pub ancestors: Vec<String>,
}

impl Default for Language {
    fn default() -> Self {
        Self {
            name: "unnamed".to_string(),
            phonemes: HashMap::new(),
            description: None,
            ancestors: vec![],
        }
    }
}

impl Language {
    pub fn description(&self) -> Option<&serde_json::Value> {
        self.description.as_ref()
    }
}
