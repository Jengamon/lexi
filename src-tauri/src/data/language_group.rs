#[cfg(test)]
mod tests;

use std::collections::{HashMap, HashSet};

use semver::Version;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::{BaseLanguage, Language, Phoneme, Protolanguage};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LanguageGroup {
    pub version: Version,
    /// Identifies the same "family" of language groups
    ///
    /// All language groups that are related to each other have the same value
    pub family_id: Uuid,
    pub protolangs: Vec<Protolanguage>,
    pub langs: Vec<Language>,
}

#[derive(Debug, thiserror::Error)]
pub enum LanguageGroupError {
    #[error("cannot create a language with a blank name")]
    LanguageEmptyName,
    #[error("the language named {0} already exists")]
    LanguageExists(String),
    #[error("cannot create a proto-language with a blank name")]
    ProtolanguageEmptyName,
    #[error("the proto-language named {0} already exists")]
    ProtolanguageExists(String),
}

// Use the crate version as the data version of the language group

// Since we use semver, we can just bake in our semver compatibility for data.
const CRATE_VERSION: &str = env!("CARGO_PKG_VERSION");

impl Default for LanguageGroup {
    fn default() -> Self {
        Self {
            version: Version::parse(CRATE_VERSION).unwrap(),
            family_id: Uuid::new_v4(),
            protolangs: vec![],
            langs: vec![],
        }
    }
}

// Adding and removing phonemes from a (proto-)language
impl LanguageGroup {
    /// Setting a phoneme
    ///
    /// The given id must either match an id in the language itself,
    /// or the id must be present in one of the languages's protolanguages.
    ///
    /// Returns success
    pub(crate) fn set_phoneme(
        protolangs: &Vec<Protolanguage>,
        lang: &mut dyn BaseLanguage,
        id: Uuid,
        phon: Phoneme,
    ) -> bool {
        if let Some(phon_slot) = lang.phonemes_mut().get_mut(&id) {
            *phon_slot = phon;
            true
        } else {
            for anc in lang.ancestors() {
                if let Some(proto) = protolangs.iter().find(|lang| lang.name() == anc) {
                    let provided_phonemes: HashSet<_> = proto.phonemes().keys().collect();
                    if provided_phonemes.contains(&id) {
                        *lang.phonemes_mut().entry(id).or_default() = phon;
                        return true;
                    }
                } else {
                    log::error!("Invalid ancestor in {}: {}", lang.name(), anc);
                }
            }
            false
        }
    }

    /// Getting a phoneme
    ///
    /// The given id must either match an id in the language itself,
    /// or be present in one of the languages's protolanguages.
    pub(crate) fn get_phoneme<'a>(
        protolangs: &'a Vec<Protolanguage>,
        lang: &'a dyn BaseLanguage,
        id: Uuid,
    ) -> Option<&'a Phoneme> {
        if let Some(phon) = lang.phonemes().get(&id) {
            Some(phon)
        } else {
            lang.ancestors()
                .iter()
                .filter_map(|anc| {
                    protolangs
                        .iter()
                        .find(|lang| lang.name() == anc && lang.phonemes().keys().any(|k| k == &id))
                })
                .find_map(|proto| proto.phonemes.get(&id))
        }
    }
}

// General changes (epoching & merging)
impl LanguageGroup {
    /// Creates a new epoch.
    ///
    /// All protolanguages are lost, and languages become protolanguages.
    /// All protolanguage base info (phonemes and morphemes) are merged into the necessary languages
    /// before it becomes a protolanguage
    pub fn epoch(&mut self) {
        // Create language phoneme lists
        let phonemes: Vec<HashMap<_, _>> = self
            .langs
            .iter()
            .map(|l| {
                l.ancestors
                    .iter()
                    .flat_map(|anc| {
                        self.protolanguage(anc)
                            .unwrap()
                            .phonemes
                            .iter()
                            .map(|(k, v)| (*k, v.clone()))
                    })
                    .collect()
            })
            .collect();
        self.protolangs.clear();
        self.protolangs = self
            .langs
            .drain(..)
            .zip(phonemes)
            .map(|(l, phons)| {
                let mut basic: Protolanguage = l.into();
                // Order matters: if both protolanguage and language specify
                // an id for a phoneme, we want to use the one in the language
                // (a language can "override" a protolanguage's phonemes)
                // This is how we implement sound change.
                basic.phonemes = phons
                    .into_iter()
                    .chain(basic.phonemes.into_iter())
                    .collect();
                basic
            })
            .collect();
    }

    /// Merges in languages of a language group.
    ///
    /// Loads only languages from an input language group, removing data that references
    /// something that isn't provided by the protolanguages in existence.
    ///
    /// ### Warning
    /// This should only be called if the input language group shares
    /// the family id with this language group. Also, is lossy in that
    /// a merge will *replace* the languages that exist.
    pub fn merge(&mut self, langs_from: &LanguageGroup) {
        let modded_langs = langs_from
            .langs
            .iter()
            .map(|lang| {
                let present_ancestors = lang
                    .ancestors
                    .iter()
                    .filter(|anc| self.protolanguage(anc).is_some())
                    .cloned()
                    .collect();
                Language {
                    ancestors: present_ancestors,
                    ..lang.clone()
                }
            })
            .collect();

        self.langs = modded_langs;
    }
}

// Language CRUD
impl LanguageGroup {
    pub fn create_language(&mut self, name: String) -> Result<(), LanguageGroupError> {
        if name.is_empty() {
            return Err(LanguageGroupError::LanguageEmptyName);
        }

        let existing = self.langs.iter().any(|lang| lang.name == name);
        if existing {
            return Err(LanguageGroupError::LanguageExists(name));
        }

        self.langs.push(Language {
            name,
            ..Default::default()
        });

        Ok(())
    }

    pub fn delete_language(&mut self, name: String) {
        let index = self.langs.iter().position(|lang| lang.name == name);

        if let Some(index) = index {
            self.langs.remove(index);
        }
    }

    pub fn language<S: AsRef<str>>(&self, name: S) -> Option<&Language> {
        self.langs.iter().find(|lang| lang.name == name.as_ref())
    }

    pub fn language_mut<S: AsRef<str>>(&mut self, name: S) -> Option<&mut Language> {
        self.langs
            .iter_mut()
            .find(|lang| lang.name == name.as_ref())
    }

    pub fn languages(&self) -> impl Iterator<Item = &Language> {
        self.langs.iter()
    }
}

// Protolanguage CRUD
impl LanguageGroup {
    pub fn create_protolanguage(&mut self, name: String) -> Result<(), LanguageGroupError> {
        if name.is_empty() {
            return Err(LanguageGroupError::ProtolanguageEmptyName);
        }

        let existing = self.protolangs.iter().any(|lang| lang.name == name);
        if existing {
            return Err(LanguageGroupError::ProtolanguageExists(name));
        }

        self.protolangs.push(Protolanguage {
            name,
            ..Default::default()
        });

        Ok(())
    }

    pub fn delete_protolanguage(&mut self, name: String) {
        let index = self.protolangs.iter().position(|lang| lang.name == name);

        if let Some(index) = index {
            self.protolangs.remove(index);
        }
    }

    pub fn protolanguage<S: AsRef<str>>(&self, name: S) -> Option<&Protolanguage> {
        self.protolangs
            .iter()
            .find(|lang| lang.name == name.as_ref())
    }

    pub fn protolanguage_mut<S: AsRef<str>>(&mut self, name: S) -> Option<&mut Protolanguage> {
        self.protolangs
            .iter_mut()
            .find(|lang| lang.name == name.as_ref())
    }

    pub fn protolanguages(&self) -> impl Iterator<Item = &Protolanguage> {
        self.protolangs.iter()
    }
}
