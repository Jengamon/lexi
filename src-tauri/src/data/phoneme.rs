use core::fmt;
use std::{borrow::Cow, collections::HashSet, fmt::Formatter};

use ipa_translate::branner_to_ipa;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Phoneme {
    pub ortho: String,
    pub primary: Phone,
    pub allo: Vec<Phone>,
}

impl Default for Phoneme {
    fn default() -> Self {
        Self {
            ortho: "".to_string(),
            primary: Phone::Null,
            allo: vec![],
        }
    }
}

#[derive(Serialize, Deserialize, Copy, Clone, Debug)]
pub enum Place {
    Bilabial,
    Labiodental,
    Dental,
    Alveolar,
    Postalveolar,
}

#[derive(Serialize, Deserialize, Hash, PartialEq, Eq, Copy, Clone, Debug)]
pub enum ObstruentAttachment {
    Ejective,
    Preaspirated,
    Aspirated,
    Breathy,
    Creaky,
}

enum ConsPhoneBase {
    Plosive,
    Fricative,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub enum Phone {
    Plosive {
        place: Place,
        voiced: bool,
        attachments: HashSet<ObstruentAttachment>,
    },
    Affricative {
        start_place: Place,
        end_place: Place,
        voiced: bool,
        attachments: HashSet<ObstruentAttachment>,
    },
    Fricative {
        place: Place,
        voiced: bool,
        attachments: HashSet<ObstruentAttachment>,
    },
    Vowel {},
    /// Null production
    Null,
}

impl Phone {
    fn get_cons_base(
        pt: ConsPhoneBase,
        place: &Place,
        voiced: bool,
        attachments: &HashSet<ObstruentAttachment>,
    ) -> String {
        branner_to_ipa(match pt {
            ConsPhoneBase::Plosive => match place {
                Place::Bilabial if attachments.contains(&ObstruentAttachment::Ejective) => "p",
                Place::Bilabial if voiced => "b",
                Place::Bilabial => "p",
                Place::Labiodental if attachments.contains(&ObstruentAttachment::Ejective) => "p[",
                Place::Labiodental if voiced => "b[",
                Place::Labiodental => "p[",
                Place::Dental if attachments.contains(&ObstruentAttachment::Ejective) => "t[",
                Place::Dental if voiced => "d[",
                Place::Dental => "t[",
                Place::Alveolar if attachments.contains(&ObstruentAttachment::Ejective) => "t",
                Place::Alveolar if voiced => "d",
                Place::Alveolar => "t",
                Place::Postalveolar if attachments.contains(&ObstruentAttachment::Ejective) => "t",
                Place::Postalveolar if voiced => "d",
                Place::Postalveolar => "t",
            },
            ConsPhoneBase::Fricative => match place {
                Place::Bilabial if attachments.contains(&ObstruentAttachment::Ejective) => "P\"",
                Place::Bilabial if voiced => "B\"",
                Place::Bilabial => "P\"",
                Place::Labiodental if attachments.contains(&ObstruentAttachment::Ejective) => "f",
                Place::Labiodental if voiced => "v",
                Place::Labiodental => "f",
                Place::Dental if attachments.contains(&ObstruentAttachment::Ejective) => "O-",
                Place::Dental if voiced => "d-",
                Place::Dental => "O-",
                Place::Alveolar if attachments.contains(&ObstruentAttachment::Ejective) => "s",
                Place::Alveolar if voiced => "z",
                Place::Alveolar => "s",
                Place::Postalveolar if attachments.contains(&ObstruentAttachment::Ejective) => "S",
                Place::Postalveolar if voiced => "3\"",
                Place::Postalveolar => "S",
            },
        })
    }

    fn aspiration<'a>(phone: &'a str, attachments: &HashSet<ObstruentAttachment>) -> Cow<'a, str> {
        if attachments.contains(&ObstruentAttachment::Preaspirated) {
            Cow::Owned(branner_to_ipa("h^") + phone)
        } else if attachments.contains(&ObstruentAttachment::Aspirated) {
            Cow::Owned(phone.to_string() + &branner_to_ipa("h^"))
        } else {
            Cow::Borrowed(phone)
        }
    }
}

impl fmt::Display for Phone {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let phone = match self {
            Phone::Null => "âˆ…".to_string(),
            Phone::Plosive {
                place,
                voiced,
                attachments,
            } => {
                let mut phone =
                    Self::get_cons_base(ConsPhoneBase::Plosive, place, *voiced, attachments);

                if attachments.contains(&ObstruentAttachment::Breathy) {
                    phone += &branner_to_ipa(r#"h")"#);
                }

                if attachments.contains(&ObstruentAttachment::Creaky) {
                    phone += &branner_to_ipa("~");
                }

                if attachments.contains(&ObstruentAttachment::Ejective) {
                    phone += &branner_to_ipa("`");
                }

                phone = Self::aspiration(&phone, attachments).to_string();

                phone
            }
            Phone::Affricative {
                start_place,
                end_place,
                voiced,
                attachments,
            } => {
                let mut start_phone =
                    Self::get_cons_base(ConsPhoneBase::Plosive, start_place, *voiced, attachments);
                let mut end_phone =
                    Self::get_cons_base(ConsPhoneBase::Fricative, end_place, *voiced, attachments);

                if attachments.contains(&ObstruentAttachment::Breathy) {
                    start_phone += &branner_to_ipa(r#"h")"#);
                    end_phone += &branner_to_ipa(r#"h")"#);
                }

                if attachments.contains(&ObstruentAttachment::Creaky) {
                    start_phone += &branner_to_ipa("~");
                    end_phone += &branner_to_ipa("~");
                }

                let mut phone = branner_to_ipa(&format!("{start_phone})){end_phone}"));

                if attachments.contains(&ObstruentAttachment::Ejective) {
                    phone += &branner_to_ipa("`");
                }

                phone = Self::aspiration(&phone, attachments).to_string();

                phone
            }
            Phone::Fricative {
                place: _,
                voiced: _,
                attachments: _,
            } => todo!(),
            Phone::Vowel {} => todo!(),
        };

        write!(f, "{}", phone)
    }
}
