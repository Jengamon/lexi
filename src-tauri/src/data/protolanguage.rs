use serde::{Deserialize, Serialize};

use super::Phoneme;

#[derive(Serialize, Deserialize, Clone)]
pub struct Protolanguage {
    pub name: String,
    pub phonemes: Vec<Phoneme>,
}

impl Default for Protolanguage {
    fn default() -> Self {
        Self {
            name: "unnamed".to_string(),
            phonemes: vec![],
        }
    }
}
