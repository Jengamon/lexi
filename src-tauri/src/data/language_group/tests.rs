use super::LanguageGroup;
use crate::data::{Language, Phone, Phoneme, Protolanguage};
use insta::{assert_yaml_snapshot, with_settings};
use std::collections::{HashMap, HashSet};
use uuid::uuid;

#[test]
fn lg_deny_set_phoneme_if_nonexistant() {
    let mut lg = LanguageGroup {
        family_id: uuid!("00000000-0000-0000-0000-000000000000"),
        protolangs: vec![Protolanguage {
            name: "1".to_string(),
            ..Default::default()
        }],
        langs: vec![Language {
            name: "1".to_string(),
            ancestors: vec!["1".to_string()],
            ..Default::default()
        }],
        ..Default::default()
    };

    assert_eq!(
        LanguageGroup::set_phoneme(
            &lg.protolangs,
            &mut lg.langs[0],
            uuid!("00000000-0000-0000-0000-000000000000"),
            Phoneme::default()
        ),
        false
    );

    lg.protolangs[0].phonemes.insert(
        uuid!("00000000-0000-0000-0000-000000000000"),
        Phoneme::default(),
    );

    assert_eq!(
        LanguageGroup::set_phoneme(
            &lg.protolangs,
            &mut lg.langs[0],
            uuid!("00000000-0000-0000-0000-000000000000"),
            Phoneme::default()
        ),
        true
    );

    with_settings!({sort_maps => true}, {
        assert_yaml_snapshot!(lg);
    })
}

#[test]
fn lg_epoch_preserves_protophonemes_in_language() {
    let mut pre = LanguageGroup {
        family_id: uuid!("00000000-0000-0000-0000-000000000000"),
        protolangs: vec![Protolanguage {
            name: "1".to_string(),
            description: None,
            phonemes: {
                let mut map = HashMap::new();
                map.insert(
                    uuid!("00000000-0000-0000-0000-000000000000"),
                    Phoneme {
                        ortho: "1".to_string(),
                        primary: Phone::Plosive {
                            place: crate::data::PlosivePlace::Bilabial,
                            voiced: false,
                            attachments: HashSet::new(),
                        },
                        allo: vec![],
                    },
                );
                map
            },
        }],
        langs: vec![Language {
            name: "1'".to_string(),
            ancestors: vec!["1".to_string()],
            description: None,
            phonemes: {
                let mut map = HashMap::new();
                map.insert(
                    uuid!("00000000-0000-0000-0000-000000000001"),
                    Phoneme {
                        ortho: "2".to_string(),
                        primary: Phone::Plosive {
                            place: crate::data::PlosivePlace::Bilabial,
                            voiced: true,
                            attachments: HashSet::new(),
                        },
                        allo: vec![],
                    },
                );
                map
            },
        }],
        ..Default::default()
    };

    pre.epoch();

    with_settings!({sort_maps => true}, {
        assert_yaml_snapshot!(pre);
    })
}

#[test]
fn epoch_when_phoneme_and_protophoneme_exist_keeps_phoneme() {
    let mut pre = LanguageGroup {
        family_id: uuid!("00000000-0000-0000-0000-000000000000"),
        protolangs: vec![Protolanguage {
            name: "1".to_string(),
            description: None,
            phonemes: {
                let mut map = HashMap::new();
                map.insert(
                    uuid!("00000000-0000-0000-0000-000000000000"),
                    Phoneme {
                        ortho: "1".to_string(),
                        primary: Phone::Plosive {
                            place: crate::data::PlosivePlace::Bilabial,
                            voiced: false,
                            attachments: HashSet::new(),
                        },
                        allo: vec![],
                    },
                );
                map
            },
        }],
        langs: vec![Language {
            name: "1'".to_string(),
            ancestors: vec!["1".to_string()],
            description: None,
            phonemes: {
                let mut map = HashMap::new();
                map.insert(
                    uuid!("00000000-0000-0000-0000-000000000000"),
                    Phoneme {
                        ortho: "2".to_string(),
                        primary: Phone::Plosive {
                            place: crate::data::PlosivePlace::Bilabial,
                            voiced: true,
                            attachments: HashSet::new(),
                        },
                        allo: vec![],
                    },
                );
                map
            },
        }],
        ..Default::default()
    };

    pre.epoch();

    with_settings!({sort_maps => true}, {
        assert_yaml_snapshot!(pre);
    })
}

#[test]
fn merge_removes_nonexistant_ancestors() {
    // TODO once we start referring to protolanguages,
    // items that reference invalid protolanguages should be wiped!
    let pre = LanguageGroup {
        family_id: uuid!("00000000-0000-0000-0000-000000000000"),
        protolangs: vec![Protolanguage {
            name: "1".to_string(),
            description: None,
            phonemes: HashMap::new(),
        }],
        langs: vec![Language {
            name: "A".to_string(),
            description: None,
            phonemes: HashMap::new(),
            ancestors: vec!["1".to_string()],
        }],
        ..Default::default()
    };

    let post_a = LanguageGroup {
        family_id: uuid!("00000000-0000-0000-0000-000000000000"),
        protolangs: vec![],
        langs: vec![Language {
            name: "2".to_string(),
            description: None,
            phonemes: HashMap::new(),
            ancestors: vec!["H".to_string()],
        }],
        ..Default::default()
    };

    let post_b = LanguageGroup {
        family_id: uuid!("00000000-0000-0000-0000-000000000000"),
        protolangs: vec![],
        langs: vec![Language {
            name: "2".to_string(),
            description: None,
            phonemes: HashMap::new(),
            ancestors: vec!["1".to_string()],
        }],
        ..Default::default()
    };

    with_settings!({sort_maps => true}, {
        assert_yaml_snapshot!({
            let mut pre = pre.clone();
            pre.merge(&post_a);
            pre
        });

        assert_yaml_snapshot!({
            let mut pre = pre.clone();
            pre.merge(&post_b);
            pre
        });
    });
}
