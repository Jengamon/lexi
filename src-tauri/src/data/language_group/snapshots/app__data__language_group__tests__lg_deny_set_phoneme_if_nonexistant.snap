---
source: src-tauri/src/data/language_group/tests.rs
expression: lg
---
version: 0.1.0
family_id: 00000000-0000-0000-0000-000000000000
protolangs:
  - name: "1"
    phonemes:
      00000000-0000-0000-0000-000000000000:
        ortho: ""
        primary: "Null"
        allo: []
    description: ~
langs:
  - name: "1"
    phonemes:
      00000000-0000-0000-0000-000000000000:
        ortho: ""
        primary: "Null"
        allo: []
    description: ~
    ancestors:
      - "1"

